File: 이다영/7. 깊이, 넓이 우선탐색/8. 사과나무.java
package main.java.org.example.inflearn.sec07;

import java.util.*;
import java.io.*;

// 사과나무
public class Inflearn7_8 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] board = new int[n][n];
        boolean[][] visited = new boolean[n][n];
        for (int i = 0; i < n; i++) {
            String[] tmp = br.readLine().split(" ");
            for (int j = 0; j < n; j++) {
                board[i][j] = Integer.parseInt(tmp[j]);
            }
        }

        int flag = 0;
        int sx = n / 2; int sy = n /2;
        int sum = board[sx][sy];
        int[] dx = new int[]{-1, 0, 1, 0};
        int[] dy = new int[]{0, 1, 0, -1};
        Deque<int[]> dq = new ArrayDeque<>();
        dq.addLast(new int[]{sx, sy});
        visited[sx][sy] = true;
        while(!dq.isEmpty()) {
            if (flag == n / 2) {
                break;
            }

            int size = dq.size();
            for (int i = 0; i < size; i++) {
                int[] tmp = dq.pollFirst();
                int x = tmp[0];
                int y = tmp[1];

                for (int j = 0; j < 4; j++) {
                    int xx = x + dx[j];
                    int yy = y + dy[j];

                    if(!visited[xx][yy]) {
                        sum += board[xx][yy];
                        visited[xx][yy] = true;
                        dq.addLast(new int[]{xx, yy});
                    }
                }
            }
            flag++;
        }

        System.out.println(sum);
    }
}

File: 박정환/7. 깊이, 넓이 우선탐색 활용/8. 사과나무.py
from collections import deque

dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]
n = int(input())
a = [list(map(int, input().split())) for _ in range(n)]
check = [[0] * n for _ in range(n)]
sum = 0
Q = deque()
check[n//2][n//2] = 1
sum += a[n//2][n//2]
Q.append((n//2, n//2))
level = 0

while True:
    if level == n//2:
        break
    size = len(Q)
    for i in range(size):
        tmp = Q.popleft()
        for j in range(4):
            x = tmp[0] + dx[j]
            y = tmp[1] + dx[j]
            if check[x][y] == 0:
                sum += a[x][y] 
                check[x][y] = 1
                Q.append((x, y))
    level += 1

print(sum)
